pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotEvaluateLiteralRule.java

```diff
@@ -0,0 +1,249 @@
+          .collect(Collectors.toList());
+      return rexCall.clone(rexCall.getType(), newOperands);
+    }
+    return rexNode;
+  }
+
+  /**
+   * Evaluates the literal only function and returns the result as a RexLiteral, null if the function cannot be
+   * evaluated.
+   */
+  protected static RexNode evaluateLiteralOnlyFunction(RexBuilder rexBuilder, RexNode rexNode,
+      Map<RexNode, RexNode> evaluatedResults) {
+    if (rexNode instanceof RexLiteral) {
+      return rexNode;
+    }
+    Preconditions.checkArgument(rexNode instanceof RexCall, "Expected RexCall, got: " + rexNode);
+    if (evaluatedResults.containsKey(rexNode)) {
+      return evaluatedResults.get(rexNode);
+    }
+    RexCall function = (RexCall) rexNode;
+    List<RexNode> operands = new ArrayList<>(function.getOperands());
+    int numOperands = operands.size();
+    for (int i = 0; i < numOperands; i++) {
+      // The RexCall is guaranteed to have all operands as RexLiteral or evaluated as RexLiteral.
+      // So recursively call evaluateLiteralOnlyFunction on all operands.
+      RexNode operand = evaluateLiteralOnlyFunction(rexBuilder, operands.get(i), evaluatedResults);
+      operands.set(i, operand);
+    }
+
+    String functionName = PinotRuleUtils.extractFunctionName(function);
+    FunctionInfo functionInfo = FunctionRegistry.getFunctionInfo(functionName, numOperands);
+    RexNode resultRexNode = rexNode;
+    if (functionInfo != null) {
+      Object[] arguments = new Object[numOperands];
+      for (int i = 0; i < numOperands; i++) {
+        RexNode operand = function.getOperands().get(i);
+        Preconditions.checkArgument(operand instanceof RexLiteral, "Expected all the operands to be RexLiteral");
+        Object value = ((RexLiteral) operand).getValue();
+        if (value instanceof NlsString) {
+          arguments[i] = ((NlsString) value).getValue();
+        } else if (value instanceof GregorianCalendar) {
+          arguments[i] = ((GregorianCalendar) value).getTimeInMillis();
+        } else if (value instanceof ByteString) {
+          arguments[i] = ((ByteString) value).getBytes();
+        } else {
+          arguments[i] = value;
+        }
+      }
+      RelDataType rexNodeType = rexNode.getType();
+      Object functionResult;
+      try {
+        FunctionInvoker invoker = new FunctionInvoker(functionInfo);
+        invoker.convertTypes(arguments);
+        functionResult = invoker.invoke(arguments);
+      } catch (Exception e) {
+        throw new SqlCompilationException(
+            "Caught exception while invoking method: " + functionInfo.getMethod() + " with arguments: "
+                + Arrays.toString(arguments), e);
+      }
+      if (functionResult == null) {
+        resultRexNode = rexBuilder.makeNullLiteral(rexNodeType);
+      } else if (rexNodeType.getSqlTypeName() == SqlTypeName.TIMESTAMP) {
+        long millis = (functionResult instanceof Number) ? ((Number) functionResult).longValue()
```

Comment:
Also handle `Timestamp` type explicitly. It should be the most common case
```suggestion
        long millis;
        if (functionResult instanceof Timestamp) {
          millis = ((Timestamp) functionResult).getTime();
        } else if (functionResult instanceof Number) {
          millis = ((Number) functionResult).longValue();
        } else {
          millis = TimestampUtils.toMillisSinceEpoch(functionResult.toString());
        }
```